File: 1250-longest-common-subsequence/1250-longest-common-subsequence.java
class Solution {
    private int find(String text1, String text2, int id1, int id2, int[][] dp){
        if(id1<0 || id2<0) return 0;

        if(dp[id1][id2] != -1) return dp[id1][id2];

        if(text1.charAt(id1) == text2.charAt(id2)){
            return dp[id1][id2] = 1 + find(text1, text2, id1-1, id2-1, dp);
        }

        return dp[id1][id2] = Math.max(find(text1, text2, id1-1, id2, dp), 
                                        find(text1, text2, id1, id2-1, dp));
    }

    public int longestCommonSubsequence(String text1, String text2) {
        int m = text1.length();
        int n = text2.length();

        int[][] dp = new int[m][n];
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                dp[i][j] = -1;
            }
        }

        return find(text1, text2, m-1, n-1, dp);
    }
}

File: 1250-longest-common-subsequence/1250-longest-common-subsequence.py
class Solution(object):
    def longestCommonSubsequence(self, text1, text2):
        """
        :type text1: str
        :type text2: str
        :rtype: int
        """

        m = len(text1)
        n = len(text2)

        dp = [[0 for _ in range(n+1)] for _ in range(m+1)]
        
        for i in range(m-1, -1, -1):
            for j in range(n-1, -1, -1):
                same = 0
                if text1[i] == text2[j]:
                    same = 1 + dp[i+1][j+1]
                not_same = max(dp[i+1][j], dp[i][j+1])

                dp[i][j] = max(same, not_same)

        return dp[0][0]
        